<a href="https://colab.research.google.com/github/pirsquared/Advent-of-Code/blob/master/2018/Day11_improved.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Faster
`numba` isn't necessary but does help

In [1]:
!pip install numba

In [0]:
import pandas as pd
import numpy as np

from numba import njit

In [0]:
 def get_a(data):
    d = int(data)
    j = np.arange(1, 301)
    i = np.arange(1, 301)
    r = j + 10
    p = (r * i[:, None] + d) * r // 100 % 10 - 5
    return p

In [0]:
@njit
def f(c):
  n = c.shape[0]
  n = n - 1
  maxpower = np.array([0, 0, 0, 0])
  
  for i in range(n):
    for j in range(n):
      for k in range(n - max(i, j)):
        power = g(c, i, j, k)
        if power > maxpower[-1]:
          maxpower[:] = [j + 1, i + 1, k, power]
          
  return maxpower
        
@njit
def g(a, i, j, k):
  return a[i+k, j+k] - a[i, j+k] - a[i+k, j] + a[i, j]
        

# Deal Maker
By calculating the 2-D cumulative sumation, I've pre-calculated much of what will be needed to determine the convolutions.  I'll will have reduced every sum over a kernel to an operation over four numbers.

In [0]:
def pad(a):
  n, m = a.shape
  c = np.zeros((n + 1, m + 1), a.dtype)
  c[1:, 1:] = a.cumsum(0).cumsum(1)
  
  return c

In [0]:
def part2(data):
  *a, p = f(pad(get_a(int(data))))
  return f"""Power: {p}
  AoC: {','.join(map(str, a))}"""

In [7]:
print(part2(18))

Power: 113
  AoC: 90,269,16


In [8]:
print(part2(42))

Power: 119
  AoC: 232,251,12


In [9]:
print(part2(5177))

Power: 80
  AoC: 231,135,8


In [10]:
%timeit part2(5177)

10 loops, best of 3: 20.7 ms per loop


# Trying to Explain

In [0]:
import pandas as pd

In [0]:
a = get_a(18)

In [0]:
df = pd.DataFrame(a)
df_cumsum = df.cumsum(0).cumsum(1)
df_subset = df.loc[267:284, 88:105]
df_cumsum_subset = df_cumsum.loc[267:284, 88:105]

This is the 16x16 grid whos upper left corner starts at 269, 90

In [14]:
def highlight_cols(s):
    color = 'yellow'
    return f'background-color: {color}'

interior = pd.IndexSlice[268:283, 89:104]
df_subset.style.applymap(highlight_cols, subset=interior)

We can see the sum accross that array is the power we expect

In [15]:
df_subset.loc[268:283, 89:104].values.sum()

113

But instead if we calculate the 2-D cumulative sum. we can reference specific points to calculate the same as the sum the the array above.

- **Yellow** are the cumsums at same location as the yellow in the dataframe above.
- **Cyan** are the corners I will reference in the formula to calc the same sumation.

In [16]:
def highlight_cols(s):
    color = 'yellow'
    return f'background-color: {color}'
  
def highlight_cell(s):
    color = 'cyan'
    return f'background-color: {color}'

interior = pd.IndexSlice[268:283, 89:104]
four_pts = pd.IndexSlice[[267, 283], [88, 104]]
df_cumsum_subset.style \
    .applymap(highlight_cols, subset=interior) \
    .applymap(highlight_cell, subset=four_pts)


In [17]:
df_cumsum.loc[four_pts]

88     104
267 -11173 -12516
283 -11841 -13071

In [18]:
(ii, ij), (ji, jj) = df_cumsum.loc[four_pts].values
jj - ij - ji + ii

113

The point is that `jj - ij - ji + ii` is constant time while `df_subset.loc[268:283, 89:104].values.sum()` is O(k^2) time.


_____
# Another way
Slicing with numpy

In [0]:
def part2_1(data):
    c = pad(get_a(int(data)))

    i, j, k = np.indices((300, 300, 300)).reshape(3, -1)
    m = k < 300 - np.maximum(i, j)

    i = i[m]
    j = j[m]
    k = k[m]

    pos = (c[i, j] + c[i + k, j + k] - c[i, j + k] - c[i + k, j]).argmax()

    return ','.join(map(str, (j[pos] + 1, i[pos] + 1, k[pos])))

In [20]:
print(part2_1(18))

90,269,16


In [21]:
%timeit(part2_1(18))

1 loop, best of 3: 995 ms per loop
